File: provably_fair_implementation/provably_fair/ReverseHashChain.cs
using System;
using System.Security.Cryptography;

namespace provably_fair
{
    public class ReverseHashChainResult
    {
        public int index { get; }
        public byte[] seed { get; }
        public float? random { get; }

        public ReverseHashChainResult(int index, byte[] seed, float? random)
        {
            this.index = index;
            this.seed = seed;
            this.random = random;
        }
    }

    public class ReverseHashChain
    {
        private int hashSize = 100;

        private byte[] salt;
        private byte[] seed;

        public List<ReverseHashChainResult> hashList { get; }

        public ReverseHashChain(int hashSize, int maxRandomValue)
        {
            // Generate a new seed hash.
            var rng = RandomNumberGenerator.Create();

            seed = new byte[128];
            rng.GetBytes(seed);

            salt = new byte[128];
            rng.GetBytes(salt);

            // Create a reverse hash-chain.
            byte[] hash = seed;

            var hmac = new HMACSHA256(salt);
            hashList = new List<ReverseHashChainResult>();

            for (int i = 0; i < hashSize; i++)
            {
                ReverseHashChainResult result = new ReverseHashChainResult(hashSize - i, hash, GetNumberFromByteArray(hash) * maxRandomValue);
                hashList.Add(result);

                hash = hmac.ComputeHash(hash);
            }

            hashList.Reverse();
        }

        private float? GetNumberFromByteArray(byte[] hash)
        {
            var hashString = string.Join("", hash.Select(x => x.ToString("X2")));
            const int chars = 5;
            for (int i = 0; i <= hashString.Length - chars; i += chars)
            {
                var substring = hashString.Substring(i, chars);
                var number = int.Parse(substring, System.Globalization.NumberStyles.HexNumber);
                if (number > 999999)
                    continue;
                return (number % 10000) / 10000.0f;
            }
            return null;
        }
    }
}

File: provably_fair_implementation/provably_fair_kotlin/src/main/kotlin/ReverseHashChain.kt
import java.security.MessageDigest
import java.security.SecureRandom
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

data class ReverseHashChainResult(
    val index: Int,
    val seed: ByteArray,
    val random: Float?
)

class ReverseHashChain(private val hashSize: Int, private val maxRandomValue: Int) {
    val hashList: MutableList<ReverseHashChainResult> = mutableListOf()

    private val salt: ByteArray
    private val seed: ByteArray

    init {
        // Generate a new seed hash.
        val rng = SecureRandom()

        seed = ByteArray(128)
        rng.nextBytes(seed)

        salt = ByteArray(128)
        rng.nextBytes(salt)

        // Create a reverse hash-chain.
        var hash = seed

        val hmac = Mac.getInstance("HmacSHA256")
        val secretKeySpec = SecretKeySpec(salt, "HmacSHA256")
        hmac.init(secretKeySpec)

        for (i in 0 until hashSize) {
            val result = ReverseHashChainResult(hashSize - i, hash, getNumberFromByteArray(hash)?.times(maxRandomValue))
            hashList.add(result)

            hash = hmac.doFinal(hash)
        }

        hashList.reverse()
    }

    private fun getNumberFromByteArray(hash: ByteArray): Float? {
        val hashString = hash.joinToString("") { "%02X".format(it) }
        val chars = 5
        for (i in 0..hashString.length - chars step chars) {
            val substring = hashString.substring(i, i + chars)
            val number = substring.toIntOrNull(16)
            if (number != null && number <= 999999)
                return (number % 10000) / 10000.0f
        }
        return null
    }
}